runner/websocket.go

```diff
@@ -0,0 +1,131 @@
+package runner
+
+import (
+	"bytes"
+	"compress/zlib"
+	"context"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/pganalyze/collector/output/pganalyze_collector"
+	"github.com/pganalyze/collector/state"
+	"github.com/pganalyze/collector/util"
+	"golang.org/x/net/http/httpproxy"
+	"google.golang.org/protobuf/proto"
+)
+
+func SetupWebsocketForAllServers(ctx context.Context, servers []*state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
+	for idx := range servers {
+		go func(server *state.Server) {
+			logger = logger.WithPrefixAndRememberErrors(server.Config.SectionName)
+			for {
+				if server.WebSocket.Load() == nil {
+					connect(ctx, server, globalCollectionOpts, logger)
+				}
+				time.Sleep(3 * 60 * time.Second) // Delay between reconnect attempts
+			}
+		}(servers[idx])
+	}
+}
+
+func connect(ctx context.Context, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
+	connCtx, cancelConn := context.WithCancel(ctx)
+	proxyConfig := httpproxy.Config{
+		HTTPProxy:  server.Config.HTTPProxy,
+		HTTPSProxy: server.Config.HTTPSProxy,
+		NoProxy:    server.Config.NoProxy,
+	}
+	dialer := websocket.Dialer{
+		ReadBufferSize:  10240,
+		WriteBufferSize: 10240,
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			return proxyConfig.ProxyFunc()(req.URL)
+		},
+	}
+	url, _ := url.Parse(server.Config.APIBaseURL + "/v2/snapshots/websocket")
```

Comment:
To answer my own question, the error message can be confusing, but it's basically okay because of the error check below:

```
2024/09/23 16:11:55 W [server1] Error starting websocket: tls: first record does not look like a TLS handshake <nil>
2024/09/23 16:11:55 W [server2] Error starting websocket: tls: first record does not look like a TLS handshake <nil>
2024/09/23 16:12:00 E [server1] Could not collect activity for server: could not get default grant for activity snapshot: "htp://localhost:5200/v2/snapshots/grant": unsupported protocol scheme "htp"
2024/09/23 16:12:00 E [server2] Could not collect activity for server: could not get default grant for activity snapshot: "htp://localhost:5200/v2/snapshots/grant": unsupported protocol scheme "htp"
```
or
```
2024/09/23 16:13:37 W [server2] Error starting websocket: websocket: bad handshake &{400 Bad Request 400 HTTP/1.1 1 1 map[Content-Length:[166] Content-Type:[text/html] Date:[Mon, 23 Sep 2024 23:13:37 GMT] Server:[nginx/1.18.0 (Ubuntu)]] {0xc000c54060} 166 [] true false map[] 0xc0004d5b00 <nil>}
2024/09/23 16:13:37 W [server1] Error starting websocket: websocket: bad handshake &{400 Bad Request 400 HTTP/1.1 1 1 map[Content-Length:[166] Content-Type:[text/html] Date:[Mon, 23 Sep 2024 23:13:37 GMT] Server:[nginx/1.18.0 (Ubuntu)]] {0xc000e7e180} 166 [] true false map[] 0xc0001fe120 <nil>}
2024/09/23 16:13:55 E [server1] Could not collect activity for server: could not get default grant for activity snapshot: "http: //localhost:5200/v2/snapshots/grant": http: no Host in request URL
2024/09/23 16:13:55 E [server2] Could not collect activity for server: could not get default grant for activity snapshot: "http: //localhost:5200/v2/snapshots/grant": http: no Host in request URL
```

It is possible to hit a panic (I added a `%` in the host name):
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xec36af]

goroutine 107 [running]:
github.com/pganalyze/collector/runner.connect({0x18b7dc8?, 0xc0007a2000?}, 0xc000872a88, {{0xc1b49b7efae697be, 0xe90074, 0x22e87e0}, 0x1, 0x1, 0x1, 0x1, ...}, ...)
	/home/maciek/code/collector/runner/websocket.go:49 +0x24f
github.com/pganalyze/collector/runner.SetupWebsocketForAllServers.func1(0xc000872a88)
	/home/maciek/code/collector/runner/websocket.go:26 +0x1ab
created by github.com/pganalyze/collector/runner.SetupWebsocketForAllServers in goroutine 1
	/home/maciek/code/collector/runner/websocket.go:22 +0x8c
```